bus-mapping/src/state_db.rs

```diff
@@ -51,6 +42,56 @@ impl CodeDB {
     pub fn empty_code_hash() -> Hash {
         *EMPTY_CODE_HASH
     }
+
+    /// Compute number of rows required for bytecode table.
+    pub fn num_rows_required_for_bytecode_table(&self) -> usize {
+        self.0.values().map(|bytecode| bytecode.len() + 1).sum()
+    }
+
+    /// Query code in raw bytes by H256
+    pub fn get(&self, codehash: &H256) -> Option<&Vec<u8>> {
+        self.0.get(codehash)
+    }
+
+    /// Query Bytecode by H256
+    pub fn get_bytecode(&self, codehash: &H256) -> Option<Bytecode> {
+        self.0.get(codehash).cloned().map(|code| code.into())
+    }
+
+    /// Query Bytecode by U256
+    pub fn get_from_word(&self, codehash: &Word) -> Option<Bytecode> {
+        self.get_bytecode(&H256::from_uint(codehash))
+    }
+
+    /// Get raw bytes
+    pub fn to_raw(&self) -> Vec<Vec<u8>> {
```

Comment:
I feel this is a bad API. We shouldn't expose raw bytes for people to manipulate.
It is currently used in two places.
- One is in the keccak circuit, which takes `Vec<Vec<u8>>`. I didn't address keccak circuit in this pr to limit the scope.
- The other one is BytecodeCircuitAssignment. We need the raw bytes to preserve bytecode order for the test cases.

Reply:
I think there is option also avoiding refactor pretty much for other part is implementing iterator trait on CodeDB and iterate through &byte (need to check whether at the caller side, still need byte copy or not)

Reply:
I realized CodeDB has into_iterator trait. I use it to replace to_raw in [bcbef94](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1529/commits/bcbef9417b78d3245b3d41de75865d105c6ae699)